src/Microsoft.AspNetCore.Mvc.Core/Formatters/OutputFormatter.cs

```diff
@@ -76,9 +93,9 @@ protected virtual bool CanWriteType(Type type)
                         mediaTypes.Add(mediaType);
                     }
                 }
-
-                return mediaTypes;
             }
+
+            return mediaTypes;
```

Comment:
Given that `[Produces]` (or more generally, `ObjectResultExecutor`) will only support concrete media types for `contentType`, I think we already satisfy the spec you've described. It's a little clearer in the updated implementation, because it no longer relies on the trick of replacing `contentType` with `*/*` when we want to output all concrete media types.

> If contentType is non-null ... It should be a concrete media type ... should be returned if it's a match for one of the concrete media types or ... one of the wildcards

There's a bit of extra subtlety around parameters. When matching a concrete media type, we have historically not allowed `contentType` to specify extra params, and are retaining that behaviour. Whereas when matching a wildcard media type, we do want to allow `contentType` to have extra params because that's a requested feature, so we do so. In that sense, the notion of "concrete" can be misleading, because it still represents a set rather than just a single literal string.

> If contentType is non-null then it came from a produces. It should be a concrete media type.

Anyway, it's a good job you raised this point because, although there is logic in `ProducesAttribute.cs` and `ObjectResultExecutor.cs` to enforce concreteness, I hadn't yet updated it to recognize that new formats like `application/*+json` are not concrete. I've now modified the tests and logic (well, simplified it to use `.HasWildcard`) in those files.